File: bot.js
const { Client, GatewayIntentBits } = require('discord.js');

const DISCORD_TOKEN = process.env.DISCORD_TOKEN;
const STATS_URL = process.env.STATS_URL || 'https://messenger-server-klav.onrender.com/api/client/stats';
const PREFIX = process.env.BOT_PREFIX || '!';

async function getFetch() {
  if (typeof fetch === 'function') return fetch;
  const mod = await import('node-fetch');
  return mod.default;
}

if (!DISCORD_TOKEN) {
  console.error('Missing DISCORD_TOKEN env var');
  process.exit(1);
}

async function fetchStats() {
  const fetchImpl = await getFetch();
  const res = await fetchImpl(STATS_URL, { method: 'GET' });
  if (!res.ok) {
    throw new Error(`HTTP ${res.status}`);
  }
  return await res.json();
}

function formatTs(ts) {
  if (!ts) return '—';
  try {
    const d = new Date(ts);
    return d.toLocaleString('ru-RU');
  } catch {
    return String(ts);
  }
}

const client = new Client({
  intents: [GatewayIntentBits.Guilds, GatewayIntentBits.GuildMessages, GatewayIntentBits.MessageContent]
});

client.once('ready', () => {
  console.log(`Bot logged in as ${client.user.tag}`);
  console.log(`Using stats url: ${STATS_URL}`);
});

client.on('messageCreate', async (message) => {
  if (message.author.bot) return;
  if (!message.content.startsWith(PREFIX)) return;

  const [cmd] = message.content.slice(PREFIX.length).trim().split(/\s+/);
  if (!cmd) return;

  if (cmd.toLowerCase() !== 'stats' && cmd.toLowerCase() !== 'online') return;

  try {
    const data = await fetchStats();
    const online = data?.online ?? 0;
    const totalStarts = data?.stats?.totalStarts ?? 0;
    const totalStops = data?.stats?.totalStops ?? 0;
    const lastStartAt = formatTs(data?.stats?.lastStartAt);
    const lastStopAt = formatTs(data?.stats?.lastStopAt);
    const ttlSec = Math.round((data?.ttlMs ?? 0) / 1000);

    await message.reply(
      `AeroProject stats\n` +
      `Online clients: ${online} (ttl ${ttlSec}s)\n` +
      `Total starts: ${totalStarts}\n` +
      `Total stops: ${totalStops}\n` +
      `Last start: ${lastStartAt}\n` +
      `Last stop: ${lastStopAt}`
    );
  } catch (e) {
    await message.reply(`Failed to fetch stats: ${e.message || e}`);
  }
});

client.login(DISCORD_TOKEN);

File: bot.py
import os
import asyncio
from datetime import datetime

import aiohttp
import discord
from discord.ext import commands

DISCORD_TOKEN = os.getenv("DISCORD_TOKEN")
PREFIX = os.getenv("BOT_PREFIX", "!")

API_BASE_URL = os.getenv("API_BASE_URL")
CLIENT_STATS_URL = os.getenv("CLIENT_STATS_URL")

if not DISCORD_TOKEN:
    raise SystemExit("Missing DISCORD_TOKEN env var")


def _resolve_stats_url() -> str:
    if CLIENT_STATS_URL:
        return CLIENT_STATS_URL
    if API_BASE_URL:
        return API_BASE_URL.rstrip("/") + "/api/client/stats"
    return "https://botdisxc.onrender.com/api/client/stats"


STATS_URL = _resolve_stats_url()


def _format_ts_ms(ts_ms):
    if not ts_ms:
        return "—"
    try:
        dt = datetime.fromtimestamp(float(ts_ms) / 1000.0)
        return dt.strftime("%d.%m.%Y %H:%M:%S")
    except Exception:
        return str(ts_ms)


async def fetch_stats(session: aiohttp.ClientSession):
    async with session.get(STATS_URL) as resp:
        if resp.status != 200:
            raise RuntimeError(f"HTTP {resp.status}")
        return await resp.json()


intents = discord.Intents.default()
intents.guilds = True
intents.messages = True
intents.message_content = True

bot = commands.Bot(command_prefix=PREFIX, intents=intents)


@bot.event
async def on_ready():
    print(f"Bot logged in as {bot.user}")
    print(f"Using stats url: {STATS_URL}")


async def _reply_stats(ctx: commands.Context):
    async with aiohttp.ClientSession() as session:
        data = await fetch_stats(session)
        online = data.get("online", 0)
        ttl_ms = data.get("ttlMs", 0) or 0
        ttl_sec = round(ttl_ms / 1000)

        stats = data.get("stats") or {}
        total_starts = stats.get("totalStarts", 0)
        total_stops = stats.get("totalStops", 0)
        last_start_at = _format_ts_ms(stats.get("lastStartAt"))
        last_stop_at = _format_ts_ms(stats.get("lastStopAt"))

        await ctx.reply(
            "AeroProject client stats\n"
            f"Online clients: {online} (ttl {ttl_sec}s)\n"
            f"Total starts: {total_starts}\n"
            f"Total stops: {total_stops}\n"
            f"Last start: {last_start_at}\n"
            f"Last stop: {last_stop_at}"
        )


@bot.command(name="stats")
async def stats_cmd(ctx: commands.Context):
    try:
        await _reply_stats(ctx)
    except Exception as e:
        await ctx.reply(f"Failed to fetch stats: {e}")


@bot.command(name="online")
async def online_cmd(ctx: commands.Context):
    await stats_cmd(ctx)


async def main():
    async with bot:
        await bot.start(DISCORD_TOKEN)


if __name__ == "__main__":
    asyncio.run(main())
